### PR TITLE
Update form instance models and backend logic to support completing a form

### PR DESCRIPTION
**Pull request by kaiyangzheng**
## Description/Problem

Closes [story](https://trello.com/c/SesxgMJX/88-update-models-to-support-completing-a-form)

We want to know when forms are completed (all signatures signed), and give users the ability to mark a form as completed (manual verification to fully "complete" the form). 

## Solution

Updated models to add markedAsCompleted, completedAt, and markedAsCompletedAt fields. Added logic to form instance services to support new fields.

## Dependencies

[ ] This PR adds new dependencies

## Testing

Manual tests: signing signatures and checking form is completed, marking form as completed and checking form is marked as complete

### CODE REVIEW

File: apps/server/prisma/migrations/20231119014249_test/migration.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+-- AlterTable
+ALTER TABLE "FormInstance" ADD COLUMN     "completedAt" TIMESTAMP(3),
```
Comment (elvincheng3) [inline apps/server/prisma/migrations/20231119014249_test/migration.sql:2]: we might want to rename this migration

File: apps/server/prisma/seed.ts [hunk 2]
```diff
@@ -23,13 +23,13 @@ const MANAGER_SIG_FIELD_UUID = '1727a4d2-b22c-42de-b63d-5f553e964d75';
...
 
 // type definition for employee data used in upsertEmployee
-type EmployeeData = {
-  id: string;
-  firstName: string;
-  lastName: string;
-  email: string;
-  positionId: string;
-};
+// type EmployeeData = {
```
Comment (elvincheng3) [inline apps/server/prisma/seed.ts:26]: we're not using this type?

File: apps/server/src/form-instances/entities/form-instance.entity.ts [hunk 3]
```diff
@@ -18,12 +18,21 @@ export class FormInstanceEntity implements FormInstance {
...
+  @ApiProperty()
+  markedCompleted: boolean;
+
   @ApiProperty()
   createdAt: Date;
 
   @ApiProperty()
   updatedAt: Date;
 
+  @ApiProperty()
```
Comment (elvincheng3) [inline apps/server/src/form-instances/entities/form-instance.entity.ts:30]: would suggest adding a property:
```
@ApiProperty({
    required: false
})
```

File: apps/server/src/form-instances/form-instances.service.ts [hunk 4]
```diff
@@ -300,6 +300,42 @@ export class FormInstancesService {
...
+    const allSigned = formInstance.signatures.every((sig) => sig.signed);
+
+    if (allSigned) {
+      await this.prisma.formInstance.update({
+        where: { id: formInstanceId },
+        data: { completed: true, completedAt: new Date() },
+      });
+    }
+
+    const updatedFormInstance = await this.findOne(formInstanceId);
```
Comment (elvincheng3) [inline apps/server/src/form-instances/form-instances.service.ts:312]: the update method should return the updated form instance, so we wouldn't have to requery

File: apps/server/src/form-instances/form-instances.service.ts [hunk 5]
```diff
@@ -300,6 +300,42 @@ export class FormInstancesService {
...
+
+    return updatedFormInstance;
+  }
+
+  async markFormInstanceAsCompleted(formInstanceId: string) {
+    const formInstance = await this.prisma.formInstance.findUnique({
+      where: { id: formInstanceId },
+    });
+
+    if (!formInstance) {
```
Comment (elvincheng3) [inline apps/server/src/form-instances/form-instances.service.ts:328]: we can also consider adding handling in cases where the form instance is already marked completed, and when the form instance is not ready to be marked as completed

File: apps/server/src/form-instances/form-instances.service.ts [hunk 6]
```diff
@@ -300,6 +300,42 @@ export class FormInstancesService {
...
+        FormInstanceErrorMessage.FORM_INSTANCE_NOT_FOUND,
+      );
+    }
+
+    const updatedFormInstance = await this.prisma.formInstance.update({
+      where: { id: formInstanceId },
+      data: { markedCompleted: true, markedCompletedAt: new Date() },
+    });
+
+    return updatedFormInstance;
```
Comment (elvincheng3) [inline apps/server/src/form-instances/form-instances.service.ts:339]: let's also make a ticket for adding tests for these new conditions and methods? i'd link the ticket onto this comment for visibility, and leave the ticket under backend backlog

File: apps/server/prisma/seed.ts (hunk 2 above)
Comment (kaiyangzheng) [reply apps/server/prisma/seed.ts:26]: it was giving me a prettier error so I commented it out while trying to fix something in seed

File: apps/server/src/form-instances/form-instances.service.ts (hunk 4 above)
Comment (kaiyangzheng) [reply apps/server/src/form-instances/form-instances.service.ts:312]: i needed to make this query so the updated form instance would also reflect the updated signatures regardless of allSigned. I cleaned the code up a little in the latest commit

File: apps/server/src/form-instances/form-instances.service.ts (hunk 6 above)
Comment (kaiyangzheng) [reply apps/server/src/form-instances/form-instances.service.ts:335]: https://trello.com/c/c8i6zWCj/116-new-form-instances-tests

**Review by elvincheng3 — APPROVED** [review]

**Comment by elvincheng3** [comment]
[update models to support "completing" a form](https://trello.com/c/SesxgMJX/88-update-models-to-support-completing-a-form)

**Pull request merged**